### DISCUSSION TITLE
Upgrading Container

### DISCUSSION THREAD

**Discussion by Kameechewa** [wolveix/satisfactory-server#318] [General]
I am using Portainer to manage my containers as I'm new to docker and thought this would be easier for me to start with. I am a little confused on updating the container. I tried updating by clicking "Recreate" in Portainer; which appears to have worked as "docker pull wolveix/satisfactory-server:latest" reports it is already up to date. However, in environment variables VERSION is still 1.8.4 which was the latest version when I created the container. So, does this environment variable really mean or do anything? Should I be manually changing this for each update? If it pulled the latest version I assume it has to be up to date regardless of what VERSION is, right?

**Comment by wolveix** [comment]
Hi @Kameechewa, I don't use Portainer personally so your mileage may vary. No, you should not update the `VERSION` environment variable manually, this is pre-populated by image. For whatever reason, you're not pulling the latest version as it's currently `v1.8.5`.

From a very quick Google, it looks like there's a setting within Portainer that may force the image to be re-pulled on start.

**Reply by Kameechewa** [reply]
Yes, I did see that setting. I do have "Always pull the image" checked on. I tried both Recreate and Duplicate/Edit, both leave the version as 1.8.4. But also running "docker pull wolveix/satisfactory-server:latest" returns with:

```
latest: Pulling from wolveix/satisfactory-server
Digest: sha256:8a968aa693716ca88da8680409c54f2f3daff71a76fb2ac2c990a782673a81e8
Status: Image is up to date for wolveix/satisfactory-server:latest
docker.io/wolveix/satisfactory-server:latest
```

If I manually click on the image. Portainer does indeed say it is version 1.8.5. But, when looking at the container variables it still shows 1.8.4.

Perhaps I need to post in a Portainer forum?

![image](https://github.com/user-attachments/assets/cefbe23f-3a65-464b-9a68-3fe19ef53c3b)
![image](https://github.com/user-attachments/assets/e0c5aad3-f8c7-4dab-acfa-6e4e5440afc8)
![image](https://github.com/user-attachments/assets/cd0e39e6-71df-4863-aab7-db9c8e196658)
![image](https://github.com/user-attachments/assets/63b8913f-8c14-4c6d-8df1-13dae91a5d92)
![image](https://github.com/user-attachments/assets/b5f8ba12-046b-4a2d-bd44-bbd37a0a6181)

**Reply by wolveix** [reply]
@Kameechewa could you send the first few lines of your container's log, please?

**Reply by Kameechewa** [reply]
```
===== Satisfactory Server 1.8.4 =====
https://github.com/wolveix/satisfactory-server
WARNING: VMOVERRIDE is enabled, skipping CPU model check. Satisfactory might crash!
Checking available memory: 11GB detected
WARNING: You have less than the required 12GB minmum (11GB detected) of available RAM to run the game server.
It is likely that the server will fail to load properly.
Clearing old Satisfactory logs (set LOG=true to disable this)
Setting autosave number to 5
Setting max objects to 2162688
Setting max tick rate to 30
Setting server streaming to 1
Setting timeout to 30
Setting max players to 4
Checking available storage: 35GB detected
Downloading the latest version of the game...
Redirecting stderr to '/home/steam/.steam/steam/logs/stderr.txt'
ILocalize::AddFile() failed to load file "public/steambootstrapper_english.txt".
```

Do you need more? I do see that is shows 1.8.4 up top but Portainer swears I'm using the 1.8.5 image.

**Reply by wolveix** [reply]
In your screenshot showing the image labels, the image has is wrong, which would explain it being 1.8.4, despite Portainer also telling you that it had pulled the latest image (which does show the correct hash). My guess is that Portainer isn't actually recreating your container with the latest image, or you're somehow restarting the container instead of recreating it. I see there is an option to delete the image (with the incorrect hash), perhaps see whether that helps?

**Reply by Kameechewa** [reply]
Sorry if I'm missing something obvious but aren't all the hashes in my screenshots the same? I tried deleting any images pertaining to satisfactory. This is what I have after manually downloading the image:

![image](https://github.com/user-attachments/assets/9e2bd2ed-2192-4011-bfc6-cdc2a6c1c9f7)

So, according to Portainer, and docker, I should have the latest and that should be the only image for it.

```
REPOSITORY                     TAG       IMAGE ID       CREATED        SIZE
wolveix/satisfactory-server    latest    3e60f0d08cd3   31 hours ago   352MB
portainer/portainer-ce         2.21.1    9c0b5a6a7e4d   6 days ago     301MB
homeassistant/home-assistant   stable    1d2978a65a28   9 days ago     1.76GB
itzg/minecraft-server          latest    34354a1ccef3   4 months ago   942MB
```

Looking at the container itself it should be correct:

```
CONTAINER ID   IMAGE                                 COMMAND        CREATED             STATUS             PORTS                                                                                            NAMES
5f75ef66e7e2   wolveix/satisfactory-server:latest    "/init.sh"     4 minutes ago       Up 4 minutes         

sudo docker container ls --all --filter=ancestor=3e60f0d08cd3 --format "{{.ID}}"
5f75ef66e7e2
```

I event tried creating a new container and it still reports 1.8.4. Now I could be completely wrong here and I'm definitely not trying to say you did anything wrong as I know nothing about making images, but is it possible that the version numbers need updated somewhere in the image?

Looking here it also seems to indicate it's using the right image, I think?
![image](https://github.com/user-attachments/assets/a60ffb79-82c8-426e-b3e0-9939d4f8aed9)

**Reply by wolveix** [reply]
@Kameechewa no, the VERSION is correctly set in the published image.

In the message you wrote before pasting the screenshots, you sent this:

> latest: Pulling from wolveix/satisfactory-server
> Digest: sha256:8a968aa693716ca88da8680409c54f2f3daff71a76fb2ac2c990a782673a81e8
> Status: Image is up to date for wolveix/satisfactory-server:latest
> docker.io/wolveix/satisfactory-server:latest

That is the correct hash. In every single reference to the image hash in your screenshots, the hash is wrong. Either you're not actually recreating the container, or something else is going on. Per my previous message, did you try deleting the old version using the button shown in your screenshots?

**Reply by Kameechewa** [reply]
Thank you, I understand what you are saying now. After a lot of trial and error I had to completely delete the container and start it with `sudo docker run -d --name satisfactory --restart unless-stopped wolveix/satisfactory-server@sha256:8a968aa693716ca88da8680409c54f2f3daff71a76fb2ac2c990a782673a81e8`. Now in the logs it shows as running version 1.8.5 but for some reason Portainer still shows the 1.8.4 hash as does "docker images" but Portainer shows the version as 1.8.5. The config for the container also shows the old hash. So how is it running with 1.8.5 I have no idea.

I think I need to post on the Portainer forum.

[Edit]: Running "sudo docker inspect wolveix/satisfactory-server:latest" returns the following (again, I know nothing about images or docker really):

```
[
    {
        "Id": "sha256:3e60f0d08cd3e3949bfa631f35edc61aa6787c56009a5d3928e585498e06f6b7",
        "RepoTags": [
            "wolveix/satisfactory-server:latest"
        ],
        "RepoDigests": [
            "wolveix/satisfactory-server@sha256:8a968aa693716ca88da8680409c54f2f3daff71a76fb2ac2c990a782673a81e8"
```

So it seems like Portainer is showing the ID whereas docker was showing the digest? I'm also reading about how sometimes it only grabs the changed layers so it may not grab the new info? Again, don't feel like you need to explain any of that to me, I likely won't understand anyways. Just trying to help I guess.

**Reply by wolveix** [reply]
Is it possible that Portainer persists the environment variables between recreations? I'm wondering if it ignores the image's preset value(s) in favour of its own. That would make sense for overridden values, but not otherwise. This is why I'm not personally a big fan of Portainer :) While it adds a nice interface, it also adds a lot of confusion
